packages/netlify/src/index.ts

```diff
@@ -1,2 +1,220 @@
+				})});
+				export const config = { name: "Astro SSR", generator: "@astrojs/netlify@${packageVersion}", path: "/*", preferStatic: true };
+			`
+		);
+	}
+
+	async function writeMiddleware(entrypoint: URL) {
+		await mkdir(middlewareOutputDir(), { recursive: true });
+		await writeFile(
+			new URL('./entry.mjs', middlewareOutputDir()),
+			`
+			import { onRequest } from "${fileURLToPath(entrypoint).replaceAll('\\', '/')}";
+			import { createContext, trySerializeLocals } from 'astro/middleware';
+
+			export default async (request, context) => {
+				const ctx = createContext({ 
+					request,
+					params: {}
+				});
+				ctx.locals = { netlify: { context } }
+				const next = () => {
+					const { netlify, ...otherLocals } = ctx.locals;
+					request.headers.set("x-astro-locals", trySerializeLocals(otherLocals));
+					return context.next();
+				};
+			
+				return onRequest(ctx, next);
+			}
+
+			export const config = {
+				name: "Astro Middleware",
+				generator: "@astrojs/netlify@${packageVersion}",
+				path: "/*", excludedPath: ["/_astro/*", "/.netlify/images/*"]
+			};
+			`
+		);
+
+		// taking over bundling, because Netlify bundling trips over NPM modules
+		await build({
+			entryPoints: [fileURLToPath(new URL('./entry.mjs', middlewareOutputDir()))],
+			target: 'es2022',
+			platform: 'neutral',
+			outfile: fileURLToPath(new URL('./middleware.mjs', middlewareOutputDir())),
+			allowOverwrite: true,
+			format: 'esm',
+			bundle: true,
+			minify: false,
+		});
+	}
+
+	return {
+		name: '@astrojs/netlify',
+		hooks: {
+			'astro:config:setup': async ({ config, updateConfig }) => {
+				rootDir = config.root;
+				await cleanFunctions();
+
+				outDir = new URL('./dist/', rootDir);
+
+				const isRunningInNetlify = Boolean(
+					process.env.NETLIFY || process.env.NETLIFY_LOCAL || process.env.NETLIFY_DEV
+				);
+
+				updateConfig({
+					outDir,
+					build: {
+						redirects: false,
+						client: outDir,
+						server: ssrOutputDir(),
+					},
+					vite: {
+						server: {
+							watch: {
+								ignored: [fileURLToPath(new URL('./.netlify/**', rootDir))],
+							},
+						},
+					},
+					image: {
+						service: {
+							entrypoint: isRunningInNetlify ? '@astrojs/netlify/image-service.js' : undefined,
+						},
+					},
+				});
+			},
+			'astro:config:done': ({ config, setAdapter }) => {
+				rootDir = config.root;
+				_config = config;
+
+				if (config.image.domains.length || config.image.remotePatterns.length) {
+					throw new AstroError(
+						"config.image.domains and config.image.remotePatterns aren't supported by the Netlify adapter.",
+						'See https://github.com/withastro/adapters/tree/main/packages/netlify#image-cdn for more.'
+					);
+				}
+
+				setAdapter({
+					name: '@astrojs/netlify',
+					serverEntrypoint: '@astrojs/netlify/ssr-function.js',
+					exports: ['default'],
+					adapterFeatures: {
+						functionPerRoute: false,
+						edgeMiddleware: true,
+					},
```

Comment:
The removal of `functionPerRoute` is explained here: https://github.com/withastro/adapters/pull/84/files#diff-61372ae322de214c90be2cb1659463fa884f96d43179cc607c9fb1cfc6abdce1R67

I've outlined why I think Middleware always needs to run at the edge here: https://github.com/withastro/adapters/pull/84#discussion_r1415353909

Reply:
> The adapterFeatures should says that the adapter supports it, right?

Could you elaborate on that? My understanding is that by saying `functionPerRoute: false` in `adapterFeatures`, Astro will generate only one big bundle, instead of one bundle for each route. That's what we want. Does `adapterFeatures` do something different than I think?

Reply:
Never mind me, I was confusing `adapterFeatures` with `supportedAstroFeatures` 🤦‍♂️